### DISCUSSION TITLE
"null" option in Radio Button

### DISCUSSION THREAD

**Discussion by mxmtsk** [filamentphp/filament#1179] [Help]
I have nullable boolean fields in my database which I wanna edit. `null` would be the default, meaning the information is "unknown.

I have setup my field the following way:

```php
Forms\Components\Radio::make('cashless_payment')
    ->label('Accepts cashless payment methods')
    ->options([
        null => 'Unknown',
        true => 'Yes',
        false => 'No',
    ]),
```

The issue I have is, that for all rows where the column is set to `null`, none of the options is selected, although I expected "Unknown" to be selected.

Am I misunderstanding something or is this a bug / by the design?

**Comment by danharrin** [comment]
Hmm, what is the HTML output of the radio button with the null value? Maybe it's trying to match empty string `''` instead.

**Reply by mxmtsk** [reply]
@danharrin The output is

```html
<input
  name="data.cashless_payment"
  id="data.cashless_payment-"
  type="radio"
  value=""
  wire:model.defer="data.cashless_payment"
  class="focus:ring-primary-500 h-4 w-4 text-primary-600 border-gray-300"
/>
```

So yes, it seems to match an empty string. Is this something I can workaround somehow? Or would you consider this a bug even?

**Reply by danharrin** [reply]
Isn't this sadly a limitation with HTML inputs? Maybe we can somehow cast `null` values to an empty string and then back to null.

```php
->afterStateHydrated(fn ($component, $state) => if (blank($state)) $component->state(''))
->dehydrateStateUsing(fn ($state) => filled($state) ? $state : null)
```

**Reply by mxmtsk** [reply]
@danharrin You're absolutely right! Your workaround works perfect for me, I can work with that. Thanks you very much!

**Reply by danharrin** [reply]
Cool. I want to add a casting system to the form builder soon which will hopefully solve this issue and others like it.

**Comment by e-m-s-y** [comment]
I was using this in Filament v3 but it seems to be broken now that I've upgraded to v4.1. Does v4 offer any alternatives to use a nullabe Radio?

**Reply by sunnnyou** [reply]
this works:
``` 
Radio::make('nullable_boolean')
    ->label('Nullable Boolean')
    ->options([
        1 => 'True',
        0 => 'False',
        2 => 'Null',
    ])
    ->stateCast(app(CustomBooleanStateCast::class))
 ```
an in CustomBooleanStateCast:
```
<?php

namespace App\Filament\Contracts;

use Filament\Schemas\Components\StateCasts\Contracts\StateCast;

class CustomBooleanStateCast implements StateCast
{
    public function __construct(
    ) {
    }

    public function get(mixed $state): mixed
    {
        return match ($state) {
            0, '0' => false,
            1, '1' => true,
            default => null,
        };
    }

    public function set(mixed $state): mixed
    {
        return match ($state) {
            false => 0,
            true => 1,
            default => 2,
        };
    }
}

```

**Reply by e-m-s-y** [reply]
@sunnnyou Thanks, can confirm that it works :)
